docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+  console.log((x as SomeInterface).name);
+}
+```
+
+✅ Narrowing with type guard
+
+```typescript
+function f(x: SomeInterface | SomeOtherInterface) {
+  if (isSomeInterface(x)) {
+    // Type of x: 'SomeInterface | SomeOtherInterface'
+    console.log(x.name); // Type of x: 'SomeInterface'. Type of x.name: 'string'.
+  }
+}
+```
+
+```typescript
+const nftMetadataResults = await Promise.allSettled(...);
+
+nftMetadataResults
+  .filter((promise) => promise.status === 'fulfilled')
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft, // Property 'value' does not exist on type 'PromiseRejectedResult'.ts(2339)
+      ...
+    ),
+  );
+```
+
+🚫 Type assertion
+
+```typescript
+(nftMetadataResults.filter(
+    (promise) => promise.status === 'fulfilled',
+  ) as { status: 'fulfilled'; value: NftUpdate }[])
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft,
+      ...
+    ),
+  );
+```
+
+✅ Use a type guard as the predicate for the filter operation, enabling TypeScript to narrow the filtered results to `PromiseFulfilledResult` at the type level
+
+```typescript
+nftMetadataResults.filter(
+    (result): result is PromiseFulfilledResult<NftUpdate> =>
+      result.status === 'fulfilled',
+  )
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft,
+      ...
+    ),
+  );
+```
+
+#### Acceptable usages of `as`
+
+Although `as` is dangerous and discouraged from being used, the following is not intended as an exhaustive list of its valid use cases.
+
+##### `as` is acceptable to use for preventing or fixing `any` usage
+
+Type assertions are unsafe, but they are still always preferred to introducing `any` into the code.
+
+- With type assertions, we still get working intellisense, autocomplete, and other IDE and compiler features using the asserted type.
+- Type assertions also provide an indication of what the author intends or expects the type to be.
+- Often, the compiler will tell us exactly what the target type for an assertion needs to be, enabling us to avoid `as any`.
+
+```typescript
+// Error: Argument of type '"getNftInformation"' is not assignable to parameter of type 'keyof NftController'.ts(2345)
+// 'getNftInformation' is a private method of class 'NftController'
+sinon.stub(nftController, 'getNftInformation');
+```
+
+🚫 `as any`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as any);
+```
+
+✅ Compiler specifies that the target type should be `keyof NftController`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as keyof typeof nftController);
+```
+
+- Even an assertion to a wrong type still allows the compiler to show us warnings and errors as the code changes, and is therefore preferrable to the dangerous radio silence enforced by `any`.
+- For type assertions to an incompatible shape, use `as unknown as` as a last resort rather than `any` or `as any`.
+
+##### `as` is always acceptable to use for TypeScript syntax other than type assertion
```

Comment:
Hmmm. This is a good callout. I found this unexpected though, because I was reading the "Acceptable usages of `as`" header for this section as "Acceptable uses of `as` type assertion", since we're in the type assertion section. This isn't an example of acceptable type assertion, it's not type assertion at all.

Maybe we could move this up to the `####` level, in a section just clarifying that this isn't at type assertion and the guidelines here do not apply.